### PR TITLE
[source-contentful] offline mode

### PR DESCRIPTION
**Pull request by Khaledgarbaya**
closes #3977

This PR adds an option to the plugin `fallbackToOffline` that allows the user to continue working on their website if they don't have internet connection

### CODE REVIEW

**Comment by Khaledgarbaya** [comment]
@pieh here is the continuation of https://github.com/gatsbyjs/gatsby/pull/3977

**Comment by pieh** [comment]
very first impression - change from env var to config might be problematic here - people would need to set it in advance - because when they would change `gatsby-config.js` when they are already offline it would cause `.cache` directory to be deleted

**Comment by Khaledgarbaya** [comment]
Ooh, that makes sense! I will revert that change

File: packages/gatsby-source-contentful/src/fetch.js [hunk 1]
```diff
@@ -1,11 +1,36 @@
...
+}
+
+module.exports = async ({ spaceId, accessToken, host, syncToken, cacheDir, environment }) => {
   // Fetch articles.
   console.time(`Fetch Contentful data`)
+
+  // If the user knows they are offline, serve them cached result
+  if (process.env.GATSBY_CONTENTFUL_OFFLINE === `true`) {
+    console.log(`Using Contentful Offline cache ⚠️`)
+    console.log(`Cache may be invalidated if you edit package.json, gatsby-node.js or gatsby-config.js files`)
```
Comment (pieh) [inline packages/gatsby-source-contentful/src/fetch.js]: Maybe those 2 lines could be moved after we read cached data, so it would would print out only if we actually can use it? Not totally sure on this, but this might be better develop experience?

File: packages/gatsby-source-contentful/src/fetch.js [hunk 2]
```diff
@@ -1,11 +1,36 @@
...
+    console.error(err)
+  }
+}
+
+module.exports = async ({ spaceId, accessToken, host, syncToken, cacheDir, environment }) => {
   // Fetch articles.
   console.time(`Fetch Contentful data`)
+
+  // If the user knows they are offline, serve them cached result
+  if (process.env.GATSBY_CONTENTFUL_OFFLINE === `true`) {
```
Comment (pieh) [inline packages/gatsby-source-contentful/src/fetch.js]: I think we should address very valid comment that is being removed below:
> // For prod builds though always fail if we can't get the latest data
and add check here to only try read cache if we are running `gatsby develop`. What do You think?

File: packages/gatsby-source-contentful/src/gatsby-node.js [hunk 3]
```diff
@@ -65,8 +65,8 @@ exports.sourceNodes = async (
     syncToken,
     spaceId,
     accessToken,
-    environment,
```
Comment (pieh) [inline packages/gatsby-source-contentful/src/gatsby-node.js]: Is this removal intentional? I think `fetchData` does need `environment` param to work correctly

File: packages/gatsby-source-contentful/src/fetch.js (hunk 2 above)
Comment (Khaledgarbaya) [reply packages/gatsby-source-contentful/src/fetch.js]: Sound good to me how do we know that `gatsby develop` was used is there a command line var set?

File: packages/gatsby-source-contentful/src/gatsby-node.js (hunk 3 above)
Comment (Khaledgarbaya) [reply packages/gatsby-source-contentful/src/gatsby-node.js]: 🙈

**Comment by Khaledgarbaya** [comment]
One more thing I think I will change the PR a bit so the offline fallback kicks in only when the server is not reachable, I can use [is-reachable](https://github.com/sindresorhus/is-reachable) library

**Comment by KyleAMathews** [comment]
Any reason you're not using touchNode? Gatsby garbage collects nodes that aren't created again but if a source plugin touches a node, it won't be GCed. See gatsby-source-drupal for an example of a plugin that uses this.

File: packages/gatsby-source-contentful/src/fetch.js [hunk 4]
```diff
@@ -19,7 +19,8 @@ module.exports = async ({ spaceId, accessToken, host, syncToken, cacheDir, envir
   console.time(`Fetch Contentful data`)
 
   // If the user knows they are offline, serve them cached result
-  if (process.env.GATSBY_CONTENTFUL_OFFLINE === `true`) {
+  // For prod builds though always fail if we can't get the latest data
+  if (process.env.GATSBY_CONTENTFUL_OFFLINE === `true` && process.env.ACTIVE_ENV === `development`) {
```
Comment (pieh) [inline packages/gatsby-source-contentful/src/fetch.js]: maybe
```diff
-  if (process.env.GATSBY_CONTENTFUL_OFFLINE === `true` && process.env.ACTIVE_ENV === `development`) {
+  if (process.env.GATSBY_CONTENTFUL_OFFLINE === `true` && process.env.NODE_ENV !== `production`) {
```
?
But yeah, we should probably make it easier to determine if we are in `build` or `develop`

**Comment by Khaledgarbaya** [comment]
@KyleAMathews oh I was not aware of that, do you mean instead of saving the data to a json file we just touch the nodes?

**Comment by KyleAMathews** [comment]
Yeah, gatsby already saves the cache. Instead you just grab all nodes created by this plugin and touch them.

**Comment by uttrasey** [comment]
Thanks for running with this @Khaledgarbaya 👍

**Comment by Khaledgarbaya** [comment]
ok @pieh @KyleAMathews I made the change to use `touchNode` instead I hope I didn't miss anything

File: packages/gatsby-source-contentful/src/gatsby-node.js [hunk 5]
```diff
@@ -42,6 +42,26 @@ exports.sourceNodes = async (
...
+
+  // If the user knows they are offline, serve them cached result
+  // For prod builds though always fail if we can't get the latest data
+  if (
+    !online &&
+    process.env.GATSBY_CONTENTFUL_OFFLINE === `true` &&
+    process.env.ACTIVE_ENV !== `production`
+  ) {
+
+    _.values(store.getState().nodes)
```
Comment (pieh) [inline packages/gatsby-source-contentful/src/gatsby-node.js]: You can use 
```diff
-_.values(store.getState().nodes)
+getNodes()
```
here

File: packages/gatsby-source-contentful/src/gatsby-node.js [hunk 6]
```diff
@@ -42,6 +42,26 @@ exports.sourceNodes = async (
...
+      .filter(n => n.internal.type === `gatsby-source-contentful`)
```
Comment (pieh) [inline packages/gatsby-source-contentful/src/gatsby-node.js]: I think You need `owner`. not `type`

File: packages/gatsby-source-contentful/src/gatsby-node.js [hunk 7]
```diff
@@ -65,8 +85,9 @@ exports.sourceNodes = async (
     syncToken,
     spaceId,
     accessToken,
-    environment,
     host,
+    environment,
+    cacheDir: `${store.getState().program.directory}/.cache`,
```
Comment (pieh) [inline packages/gatsby-source-contentful/src/gatsby-node.js]: `cacheDir` not needed anymore I think?
